typing/uniquenessAnalysis.ml

```diff
@@ -0,0 +1,557 @@
+    end
+  | Texp_array(es) ->
+      List.fold_left (fun uenv e -> check_uniqueness_exp_ e uenv) uenv es
+  | Texp_ifthenelse(if', then', else_opt) ->
+      let uenv = match is_unique_variable_opt (Some if') with
+        | Some _ -> uenv
+        | None -> check_uniqueness_exp_ if' uenv in
+      begin match else_opt with
+      | Some else' -> check_uniqueness_parallel [then'; else'] uenv
+      | None -> check_uniqueness_exp_ then' uenv
+      end
+  | Texp_sequence(e, e') ->
+      check_uniqueness_exp_ e' (check_uniqueness_exp_ e uenv)
+  | Texp_while(e, e') ->
+      without_owned (fun uenv ->
+          check_uniqueness_exp_ e' (check_uniqueness_exp_ e uenv))
+        uenv
+  | Texp_list_comprehension(e, cs) ->
+      without_owned (fun uenv ->
+          check_uniqueness_comprehensions cs (check_uniqueness_exp_ e uenv))
+        uenv
+  | Texp_arr_comprehension(e, cs) ->
+      without_owned (fun uenv ->
+          check_uniqueness_comprehensions cs (check_uniqueness_exp_ e uenv))
+        uenv
+  | Texp_for(_, _, from', to', _, body) ->
+      let uenv = check_uniqueness_exp_ from' uenv in
+      let uenv = check_uniqueness_exp_ to' uenv in
+      without_owned (check_uniqueness_exp_ body) uenv
+  | Texp_send(e, _, _, _) ->
+      check_uniqueness_exp_ e uenv
+  | Texp_new _ -> uenv
+  | Texp_instvar _ -> uenv
+  | Texp_setinstvar(_, _, _, e) ->
+      check_uniqueness_exp_ e uenv
+  | Texp_override(_, ls) ->
+      List.fold_left (fun u (_, _, e) ->
+          check_uniqueness_exp_ e u
+        ) uenv ls
+  | Texp_letmodule _ -> uenv (* TODO *)
+  | Texp_letexception(_, e) -> check_uniqueness_exp_ e uenv
+  | Texp_assert e -> check_uniqueness_exp_ e uenv
+  | Texp_lazy e -> check_uniqueness_exp_ e uenv
+  | Texp_object _ -> uenv (* TODO *)
+  | Texp_pack _ -> uenv (* TODO *)
+  | Texp_letop {let_;ands;param;body} ->
+      let uenv = mark_seen param (Seen_as param) exp uenv in
+      let uenv = check_uniqueness_binding_op let_ exp uenv in
+      let uenv = List.fold_left (fun uenv bop ->
+          check_uniqueness_binding_op bop exp uenv) uenv ands in
+      check_uniqueness_cases (OneParent param) [body] uenv
+  | Texp_unreachable -> uenv
+  | Texp_extension_constructor _ -> uenv
+  | Texp_open _ -> uenv (* TODO *)
+  | Texp_probe { handler } -> check_uniqueness_exp_ handler uenv
+  | Texp_probe_is_enabled _ -> uenv
+
+and check_uniqueness_parent exp uenv =
+  match exp.exp_desc with
+  | Texp_ident _ -> uenv
+  | Texp_tuple(xs) ->
+      List.fold_left (fun uenv e -> check_uniqueness_parent e uenv) uenv xs
+  | _ -> check_uniqueness_exp_ exp uenv
+
+and check_uniqueness_value_bindings_ vbs uenv =
+  let uenv = List.fold_left
+      (fun uenv vb -> pat_to_map vb.vb_pat (exp_to_parent vb.vb_expr) uenv)
+      uenv vbs in
+  List.fold_left (fun uenv vb -> check_uniqueness_parent vb.vb_expr uenv) uenv vbs
+
+and check_uniqueness_parallel es uenv =
+  let uenv, us = List.fold_left_map (fun u e ->
```

Comment:
Probably better to stick to using the name `uenv` rather than switching to `u` in a few places. I would bind `orig_last_seen` to `uenv.last_seen` outside this fold and use that when reseting the last_seen. That avoids the need to have multiple `uenv`s in scope at one time.